tests/unit/test_iosystem3.c

```diff
@@ -305,37 +307,39 @@ main(int argc, char **argv)
 	    if ((ret = PIOc_closefile(ncid3)))
 		ERR(ret);
 	}
-    } /* next iotype */
-
-    /* Finalize PIO systems. */
-    printf("%d pio finalizing %d\n", my_rank, even_iosysid);
-    if (even_comm != MPI_COMM_NULL)
+      } /* next iotype */
+      /* Finalize PIO systems. */
+      printf("%d pio finalizing %d\n", my_rank, even_iosysid);
+      if (even_comm != MPI_COMM_NULL)
     	if ((ret = PIOc_finalize(even_iosysid)))
     	    ERR(ret);
-    printf("%d pio finalizing %d\n", my_rank, overlap_iosysid);
-    if (overlap_comm != MPI_COMM_NULL)
-    {
+      printf("%d pio finalizing %d\n", my_rank, overlap_iosysid);
+      if (overlap_comm != MPI_COMM_NULL)
+      {
 	printf("%d calling PIOc_finalize with iosysid = %d\n", my_rank, overlap_iosysid);
 	if ((ret = PIOc_finalize(overlap_iosysid)))
 	    ERR(ret);
-    }
-    printf("%d pio finalized\n", my_rank);
-    if ((ret = PIOc_finalize(iosysid_world)))
+      }
+      printf("%d pio finalized\n", my_rank);
+      if ((ret = PIOc_finalize(iosysid_world)))
     	ERR(ret);
 
-    /* Free MPI resources used by test. */
-    if ((ret = MPI_Group_free(&overlap_group)))
+      /* Free MPI resources used by test. */
+      if ((ret = MPI_Group_free(&overlap_group)))
 	ERR(ret);
-    if ((ret = MPI_Group_free(&even_group)))
+      if ((ret = MPI_Group_free(&even_group)))
 	ERR(ret);
-    if ((ret = MPI_Group_free(&world_group)))
+      if ((ret = MPI_Group_free(&world_group)))
 	ERR(ret);
-    if (overlap_comm != MPI_COMM_NULL)
+      if (overlap_comm != MPI_COMM_NULL)
 	if ((ret = MPI_Comm_free(&overlap_comm)))
 	    ERR(ret);
-    if (even_comm != MPI_COMM_NULL)
+      if (even_comm != MPI_COMM_NULL)
     	if ((ret = MPI_Comm_free(&even_comm)))
-    	    ERR(ret);
+	  ERR(ret);
+
+    } /* my_rank < TARGET_NTASKS */
+    MPI_Barrier(MPI_COMM_WORLD);
```

Comment:
Again, why MPI_COMM_WORLD and not test_comm? Also, is this barrier really needed before finalize? If so, maybe it should be moved into the finalize function to ensure that it is always there when needed?